File: test/0020_classes.js
#!/usr/bin/env node

require('..')
const test  = require('tap').test
  , root    = require('path').resolve(__dirname, '.')
  , classes = require('bindings')({ module_root: root, bindings: 'classes' });

test("class hello world", function(t) {
  t.plan(9);

  t.type(classes.World, 'function');
  t.strictEquals(classes.World.name, 'World');
  t.type(new classes.World(), 'object');
  t.throws(classes.World, 'constructor called without "new" must throw');

  t.ok((new classes.World()) instanceof classes.World);
  t.strictEquals((new classes.World()).__proto__, classes.World.prototype);
  t.ok('greet' in classes.World.prototype);
  t.type(classes.World.prototype['greet'], 'function');
  t.ok('greet' in new classes.World());

  t.end();
});

File: test/0020_classes.cpp
#include <n2o.hpp>

using namespace n2o;

class world {
public:
    world() : greeting_("Hello") {}
    std::string greet() const { return greeting_ + " world."; }
private:
    std::string greeting_;
};

N2O_MODULE(object exports) {
    exports["World"] =
        class_<world>("World")
            . function("greet", &world::greet)
            ;
}
